### PR TITLE
FireSim x Chipyard Decoupled

### PR DESCRIPTION
**Pull request by abejgonzalez**
Sister PR: https://github.com/ucb-bar/chipyard/pull/1984

Originally FireSim and Chipyard were coupled sharing things like (1) Scala sources (and SBT sources) (2) Conda environments (3) Bridges (...) and more. However, this PR decouples FireSim from Chipyard s.t. all "target" (ie. DUT) specific features are located in Chipyard (ie. things like FireSim UART bridges, NIC bridges, etc) and FireSim provides a hook to Chipyard so that I can run the FireSim makefile itself.

This includes the following:

- Target-specific CI is moved to Chipyard
- Target-specific makefrags (the API that is exposed to Chipyard to build things with FireSim) is now moved to Chipyard (this includes the `bridges` and `firesim` makefrags). Additionally, target-specific bridges are moved to Chipyard (including all their collateral).
- FireSim has it's own copies of Scala sources that once existed in Chipyard (RocketChip, Hardfloat, Diplomacy, CDE). This also changes the `build.sbt` to depend on those sources exclusively.
- Things like the DDR configuration configs now only exist on the FireSim-side. This requires build recipes to change s.t. those configs are appended to the `PLATFORM_CONFIG` (i.e. `DDR3FRFCFSLLC4MB` needs to be added to the platform configs). This also implies that the configuration for the memory model applies to all ports coming off the target (in practice this seems to be what people do). 
- Scalafix is now updated to fix more things
- Nasti (the ugly AXI copy) now doesn't rely on `implicit p: Parameters` instead it is explicitly parameterized without implicits. This makes things more verbose but reduces the amount of globals present.
- A few specific bridges are kept in `firesim-lib` s.t. Chipyard can use them.
- `firesim-lib` now provides all APIs that Chipyard needs to build it's own bridges, build testing infra. and more. This should be minimal and be cross compilable with Chisel6 and old Chisel. Eventually this will use firrtl2.
- The `BridgeAnnotation` used to define a bridge now takes a `string` referring to a class in the `midas` SBT project. This allows decoupling of FireSim SBT sources from Chipyard sources when creating a bridge at the cost of less type safety.
- New API is added for make called `symlink_sources_and_clear` allowing downstream projects (i.e. Chipyard) to symlink sources into midas cleanly.

_Bridge APIs/Setup:_
Now that bridges can live in the target, we have APIs exposed to the target to compile a bridge. It is roughly as follows (documentation will come later) and is assuming Chipyard for this explanation.

Bridges are composed of N parts.
  * (1) Items shared between Chipyard and FireSim. This includes all case classes and IOs. These need to be fully isolated from the target (i.e. they can cross compile between Chisel3 and Chisel6, and do not depend on any target-specific classes/IOs/etc).
  * (2) The bridge stub that lives in the target. This references the items in (1) so that it can create and dump a BridgeAnnotation. This depends on `firesim-lib` which provides the necessary FIRRTL classes, scala code, etc. This is only compiled in the target.
  * (3) The bridge implementation that lives in the target *but is compiled in FireSim-only* (the target does not compile this). This is symlinked into the midas compiler so it can use `midas` APIs.
  * (4) The FireSim build system makefrag hooks (this is a set of `build/config/driver/metasim.mk` files) to which the user of FireSim needs to point to with `TARGET_PROJECT_MAKEFRAG` when invoking FireSim's make. These files copy all relevant files into midas, build the target RTL, setup configs, and more.

#### Related PRs / Issues

See above.

#### UI / API Impact

Target configs now do not reference any memory model-related configs. All memory model-related configs are to be appended to the `PLATFORM_CONFIG`.

#### Verilog / AGFI Compatibility

Should be the same. 

#### Contributor Checklist
- [ ] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [ ] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous prints/debugging code?
- [ ] Did you state the UI / API impact?
- [ ] Did you specify the Verilog / AGFI compatibility impact?

- [ ] If applicable, did you regenerate and publicly share default AGFIs?

- [ ] If applicable, did you apply the `Please Backport` label?

#### Reviewer Checklist (only modified by reviewer)
Note: to run CI on PRs from forks, comment `@Mergifyio copy main` and manage the change from the new PR.
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

#### CI Help
Add the following labels to modify the CI for a set of features.
Generally, a label added only affect subsequent changes to the PR (i.e. new commits, force pushing, closing/reopening).
See `ci:*` for full list of labels:
- `ci:fpga-deploy` - Run FPGA-based E2E testing
- `ci:local-fpga-buildbitstream-deploy` - Build local FPGA bitstreams for platforms that are released
- `ci:persist-prior-workflows` - Prevent prior CI workflows from automatically cancelling with subsequent changes
- `ci:disable` - Disable CI
- `ci:disable-scala-tests` - Disable Scala tests in CI

### CODE REVIEW

**Comment by abejgonzalez** [comment]
This PR does not work.

File: sim/midas/src/main/scala/junctions/Nasti2AXI4.scala
Comment (abejgonzalez) [inline sim/midas/src/main/scala/junctions/Nasti2AXI4.scala:34]: TODO: Revert.

File: sim/midas/src/main/scala/midas/models/dram/BankConflictModel.scala
Comment (abejgonzalez) [inline sim/midas/src/main/scala/midas/models/dram/BankConflictModel.scala:75]: TODO: Remove.

File: sim/midas/src/main/scala/midas/widgets/Widget.scala
Comment (abejgonzalez) [inline sim/midas/src/main/scala/midas/widgets/Widget.scala:149]: TODO: Revert

**Comment by abejgonzalez** [comment]
This is ready for review (it *should* pass CI tests soon)

**Review by joonho3020 — COMMENTED** [review]
Can we try linux power off on F1 since this is a pretty big PR?
On a cursory look things seems fine to me, but I would like to see it being tested before clicking approve.
And of course, I'll take some more time to actually read through things.

File: .gitmodules
Comment (joonho3020) [inline .gitmodules:28]: Stupid question: did FireSim have a hardfloat dependency?

File: deploy/util/targetprojectutils.py
Comment (joonho3020) [inline deploy/util/targetprojectutils.py:12]: So just to check my understanding...
If you run manager commands (e.g. `firesim infrasetup`), it calls into the makefiles inside chipyard to generate the target design + bridges to generate the FIRRTL & annotations, and then you call midas inside FireSim?

File: sim/firesim-lib/src/main/scala/bridges/PeekPokeBridge.scala
Comment (joonho3020) [inline sim/firesim-lib/src/main/scala/bridges/PeekPokeBridge.scala:1]: If we are operating under the assumption that all the target design specific bridges (NIC, CoSpike, BlockDev...) lives in Chipyard while bridges common to all targets (Reset, Clock, PeekPoke, FASED) lives inside FireSim, can we just remove the `firesim-lib` directory completely and place these bridges in `midas/src/main/scala/midas/widgets` ?

Comment (joonho3020) [reply sim/firesim-lib/src/main/scala/bridges/PeekPokeBridge.scala:1]: Or maybe we can move `firesim-lib` into `midas/src/main/scala/firesim-lib`?
I don't think I understand what the difference between the `firesim-lib` and `midas` directories.

Comment (joonho3020) [inline sim/firesim-lib/src/main/scala/bridges/PeekPokeBridge.scala:1]: So to my understanding, `firesim-lib` should be providing APIs to the target (CY) world for generating bridge APIs.
Can we move all these bridges inside `midas` and keep this `firesim-lib` interface minimal?

**Comment by abejgonzalez** [comment]
> Can we try linux power off on F1 since this is a pretty big PR? On a cursory look things seems fine to me, but I would like to see it being tested before clicking approve. And of course, I'll take some more time to actually read through things.

Yup. I can try that right now. I think (keyword think) this won't affect Linux boot since in the CI it runs baremetal tests (Dhrystone + rv64ui-p-simple) on Rocket + Boom based configs (which are passing).

File: .gitmodules
Comment (abejgonzalez) [reply .gitmodules:28]: This just copies how Chipyard set up it's RC SBT project. I'm not sure why Chipyard had it's own hardfloat but I figured I would keep it the same.

File: deploy/util/targetprojectutils.py
Comment (abejgonzalez) [reply deploy/util/targetprojectutils.py:12]: `firesim infrasetup` calls the `firesim/sim/Makefile`. To build RTL for Chipyard, it needs to have a reference to the makefrags in Chipyard. Those makefrags tell the FireSim `Makefile` what to do to get `.fir/.anno.json` and how to use the bridges it has. Those makefrags call the Chipyard `Makefile` directly to get the FIRRTL. So the order is:

`firesim/sim/Makefile` -> Chipyard MakeFrags -> Chipyard `Makefile` -> Got FIRRTL/ANNOs here -> Back to `firesim/sim/Makefile` to finish GG compilation.

File: sim/firesim-lib/src/main/scala/bridges/PeekPokeBridge.scala
Comment (abejgonzalez) [reply sim/firesim-lib/src/main/scala/bridges/PeekPokeBridge.scala:1]: This is as minimal as I could get it. Technically everything in that folder is needed to create `BridgeAnnotations` (which are the true API between FireSim/Chipyard).

File: deploy/util/targetprojectutils.py
Comment (joonho3020) [reply deploy/util/targetprojectutils.py:12]: Yikes

**Comment by joonho3020** [comment]
Overall, looks fine to me pending CI + f1 runs

**Review by joonho3020 — APPROVED** [review]
Approved pending CI / F1 check

**Pull request merged**
